[copy-operator] Add support for the copy operator in preparation for  making consuming and borrowing no implicit copy.

gottesmm opened the pull request:
rdar://101862423

gottesmm committed:
[copy-operator] Add support for the copy operator in preparation for making consuming and borrowing no implicit copy.

Some notes:

1. I implemented this as a contextual keyword that can only apply directly to
lvalues. This ensures that we can still call functions called copy, define
variables named copy, etc. I added tests for both the c++ and swift-syntax based
parsers to validate this. So there shouldn't be any source breaks.

2. I did a little bit of type checker work to ensure that we do not treat
copy_expr's result as an lvalue. Otherwise, one could call mutating functions on
it or assign to it, which we do not want since the result of copy_value is

3. As expected, by creating a specific expr, I was able to have much greater
control of the SILGen codegen and thus eliminate extraneous copies and other
weirdness than if we used a function and had to go through SILGenApply.

rdar://101862423

gottesmm commented:
https://github.com/apple/swift-syntax/pull/1669

@swift-ci test

kavon commented on test/SILGen/copy_expr.swift:
```diff
@@ -0,0 +1,85 @@
+// CHECK: [[COPY_BORROW:%.*]] = copy_value [[BORROW]]
+// CHECK: explicit_copy_value [[COPY_BORROW]]
+// CHECK: } // end sil function '$s9copy_expr22testCopyLoadableRValueyyF'
+func testCopyLoadableRValue() {
```
I think it's worth covering cases where the copy's result is passed to `consuming`, `borrowing`, and `inout` parameters to ensure the explicit copy is there. For example, `(copy x).prop` where `prop` has a `__consuming get`, or a regular `get`.

gottesmm replied:
@kavon I added a bunch below.

ahoppen commented on lib/Parse/ParseExpr.cpp:
```diff
@@ -422,6 +422,20 @@ ParserResult<Expr> Parser::parseExprSequenceElement(Diag<> message,
   }
 
+  if (Tok.isContextualKeyword("copy") &&
+      peekToken().isAny(tok::identifier, tok::kw_self) &&
```
As I discussed offline with @gottesmm, I think this also needs to include dollar identifiers for anonymous closure arguments.

gottesmm replied:
I fixed that.

ahoppen commented on lib/Parse/ParseExpr.cpp:
```diff
@@ -422,6 +422,20 @@ ParserResult<Expr> Parser::parseExprSequenceElement(Diag<> message,
+    SourceLoc copyLoc = consumeToken();
+    ParserResult<Expr> sub =
+        parseExprSequenceElement(diag::expected_expr_after_copy, isExprBasic);
+    if (!sub.hasCodeCompletion() && !sub.isNull()) {
+      sub = makeParserResult(new (Context) CopyExpr(copyLoc, sub.get()));
+    }
```
Why can’t we form a `CopyExpr` result if `sub` contains the code completion token?

gottesmm replied:
I don't know. I think I was just copying what we already had. Should I allow for that?

ahoppen replied:
I don’t see a reason why we shouldn’t.

gottesmm replied:
Thanks again for your help @ahoppen!

gottesmm commented:
@swift-ci smoke test

gottesmm commented:
https://github.com/apple/swift-syntax/pull/1669

@swift-ci smoke test

gottesmm committed:
[consume] Fix consume parsing so we handle code completion and dollaridentifiers correctly.

This resulted from explorations around implementing the copy expr.

rdar://109479131

gottesmm committed:
Rename MoveExpr -> ConsumeExpr to reflect the final name.

NFC.

gottesmm committed:
[consume-operator] Change consume so that its result is not considered an lvalue.

Before this change it was possible to:

1. Call mutating methods on a consume result.
2. assign into a consume (e.x.: var x = ...; (consume x) = value.

From an implementation perspective, this involved just taking the logic I
already used for the CopyExpr and reusing it for ConsumeExpr with some small
tweaks.

rdar://109479440

gottesmm committed:
[move-only] Add a test that shows the previous commit allows consume on var without errors to work.

I found that in the given case, an extra allocation was being emitted causing us
to in the success case to have an extra copy. In the previous commit, as part of
updating the codegen for consume, I also ensured that we wouldn't have that
extra allocation any more by handling the lvalue directly.

rdar://109222496

gottesmm commented:
@jckarter you probably want to take a quick additional look at this since I decided while I was here to also fix consume so it has the same behavior and also in the process fix another issue where consume was emitting proper errors for move only types, but due to an extra allocation that we introduced in the var case (due to us using rvalue logic instead of lvalue logic that the type checker work here allows us to), we would hit a copy of noncopyable type error. Now that we use the lvalue logic directly for consume like I do for copy, that is elided.

gottesmm commented:
https://github.com/apple/swift-syntax/pull/1669

@swift-ci test

gottesmm commented:
https://github.com/apple/swift-syntax/pull/1669

@swift-ci test

gottesmm commented:
I updated the tests in lldb so that they work with these changes.
